File: backend/laravel/app/Models/Patient/Patient.php
<?php

namespace App\Models\Patient;

use App\Models\Appointment\Appointment;
use App\Models\Billing\HospitalBill;
use App\Models\Hospital\Hospital;
use App\Models\Lab\LabOrder;
use App\Models\Pharmacy\Prescription;
use App\Models\System\Company;
use App\Models\System\User;
use App\Traits\BelongsToCompany;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Patient extends Model
{
    use HasFactory, BelongsToCompany;

    protected $fillable = [
        'company_id', 'hospital_id', 'patient_id', 'user_id', 'patient_type_id',
        'first_name', 'last_name', 'middle_name', 'gender', 'date_of_birth', 'age',
        'blood_group', 'marital_status', 'email', 'mobile', 'alternate_mobile',
        'address', 'city', 'state', 'country', 'zip_code',
        'guardian_name', 'guardian_relationship', 'guardian_contact', 'spouse_name',
        'father_name', 'mother_name',
        'blood_pressure', 'allergies', 'chronic_diseases', 'medications',
        'family_history', 'social_history',
        'insurance_provider', 'insurance_policy_number', 'insurance_expiry',
        'insurance_coverage', 'tpa_name', 'tpa_id',
        'emergency_contact_name', 'emergency_contact_phone', 'emergency_contact_relationship',
        'is_active', 'is_deceased', 'deceased_date', 'deceased_reason',
        'registration_date', 'registration_type', 'created_by',
    ];

    protected $casts = [
        'date_of_birth' => 'date',
        'insurance_expiry' => 'date',
        'insurance_coverage' => 'decimal:2',
        'is_active' => 'boolean',
        'is_deceased' => 'boolean',
        'deceased_date' => 'date',
        'registration_date' => 'date',
    ];

    protected static function booted(): void
    {
        static::creating(function (Patient $patient) {
            $patient->patient_id ??= 'PT-'.date('Y').'-'.strtoupper(uniqid());
        });
    }

    public function getFullNameAttribute(): string
    {
        return trim("{$this->first_name} {$this->middle_name} {$this->last_name}");
    }

    public function company(): BelongsTo
    {
        return $this->belongsTo(Company::class);
    }

    public function hospital(): BelongsTo
    {
        return $this->belongsTo(Hospital::class);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function patientType(): BelongsTo
    {
        return $this->belongsTo(PatientType::class);
    }

    public function createdBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    public function family(): HasMany
    {
        return $this->hasMany(PatientFamily::class);
    }

    public function documents(): HasMany
    {
        return $this->hasMany(PatientDocument::class);
    }

    public function visits(): HasMany
    {
        return $this->hasMany(PatientVisit::class);
    }

    public function admissions(): HasMany
    {
        return $this->hasMany(PatientAdmission::class);
    }

    public function caseSheets(): HasMany
    {
        return $this->hasMany(CaseSheet::class);
    }

    public function timeline(): HasMany
    {
        return $this->hasMany(PatientTimeline::class)->orderByDesc('event_date');
    }

    public function appointments(): HasMany
    {
        return $this->hasMany(Appointment::class);
    }

    public function bills(): HasMany
    {
        return $this->hasMany(HospitalBill::class);
    }

    public function labOrders(): HasMany
    {
        return $this->hasMany(LabOrder::class);
    }

    public function prescriptions(): HasMany
    {
        return $this->hasMany(Prescription::class);
    }
}

File: frontend/mobile-app/flutter/lib/models/patient.dart
class Patient {
  final int id;
  final String patientCode; // patients.patient_id, e.g. PT-2026-XXXX
  final String firstName;
  final String lastName;
  final String? gender;
  final DateTime? dateOfBirth;
  final int? age;
  final String? bloodGroup;
  final String? email;
  final String? mobile;
  final String? address;
  final String? allergies;
  final String? chronicDiseases;
  final String? medications;
  final String? familyHistory;
  final String? emergencyContactName;
  final String? emergencyContactPhone;

  const Patient({
    required this.id,
    required this.patientCode,
    required this.firstName,
    required this.lastName,
    this.gender,
    this.dateOfBirth,
    this.age,
    this.bloodGroup,
    this.email,
    this.mobile,
    this.address,
    this.allergies,
    this.chronicDiseases,
    this.medications,
    this.familyHistory,
    this.emergencyContactName,
    this.emergencyContactPhone,
  });

  String get fullName => '$firstName $lastName'.trim();

  factory Patient.fromJson(Map<String, dynamic> json) {
    return Patient(
      id: (json['id'] as num).toInt(),
      patientCode: (json['patient_id'] ?? '').toString(),
      firstName: (json['first_name'] ?? '').toString(),
      lastName: (json['last_name'] ?? '').toString(),
      gender: json['gender'] as String?,
      dateOfBirth: _parseDate(json['date_of_birth']),
      age: (json['age'] as num?)?.toInt(),
      bloodGroup: json['blood_group'] as String?,
      email: json['email'] as String?,
      mobile: json['mobile'] as String?,
      address: json['address'] as String?,
      allergies: json['allergies'] as String?,
      chronicDiseases: json['chronic_diseases'] as String?,
      medications: json['medications'] as String?,
      familyHistory: json['family_history'] as String?,
      emergencyContactName: json['emergency_contact_name'] as String?,
      emergencyContactPhone: json['emergency_contact_phone'] as String?,
    );
  }
}

DateTime? _parseDate(dynamic value) {
  if (value == null) return null;
  return DateTime.tryParse(value.toString());
}

class PatientTimelineEvent {
  final int id;
  final String eventType;
  final String? description;
  final DateTime? eventDate;

  const PatientTimelineEvent({
    required this.id,
    required this.eventType,
    this.description,
    this.eventDate,
  });

  factory PatientTimelineEvent.fromJson(Map<String, dynamic> json) {
    return PatientTimelineEvent(
      id: (json['id'] as num).toInt(),
      eventType: (json['event_type'] ?? 'event').toString(),
      description: json['description'] as String?,
      eventDate: _parseDate(json['event_date']),
    );
  }
}
